awswrangler/lakeformation/_read.py

```diff
@@ -0,0 +1,250 @@
+            "date_as_object": True,
+            "ignore_metadata": True,
+            "strings_to_categorical": False,
+            "categories": categories,
+            "safe": safe,
+            "types_mapper": _data_types.pyarrow2pandas_extension,
+        }
+    df: pd.DataFrame = _utils.ensure_df_is_mutable(df=table.to_pandas(**args))
+    return df
+
+
+def _resolve_sql_query(
+    query_id: str,
+    chunked: Optional[bool],
+    categories: Optional[List[str]],
+    safe: bool,
+    use_threads: bool,
+    boto3_session: boto3.Session,
+) -> Union[pd.DataFrame, Iterator[pd.DataFrame]]:
+    client_lakeformation: boto3.client = _utils.client(service_name="lakeformation", session=boto3_session)
+
+    wait_query(query_id=query_id, boto3_session=boto3_session)
+
+    # The LF Query Engine distributes the load across workers
+    # Retrieve the tokens and their associated work units until NextToken is ''
+    # One Token can span multiple work units
+    # PageSize determines the size of the "Units" array in each call
+    # TODO: Inquire about good page size # pylint: disable=W0511
+    scan_kwargs: Dict[str, Union[str, int]] = {"QueryId": query_id, "PageSize": 10}
+    next_token: str = "init_token"  # Dummy token
+    token_work_units: List[Tuple[str, int]] = []
+    while next_token:
+        response = client_lakeformation.get_work_units(**scan_kwargs)
+        token_work_units.extend(  # [(Token0, WorkUnitId0), (Token0, WorkUnitId1), (Token1, WorkUnitId2) ... ]
+            [
+                (unit["Token"], unit_id)
+                for unit in response["Units"]
+                for unit_id in range(unit["WorkUnitIdMin"], unit["WorkUnitIdMax"] + 1)  # Max is inclusive
+            ]
+        )
+        next_token = response.get("NextToken", None)
+        scan_kwargs["NextToken"] = next_token
+
+    dfs: List[pd.DataFrame] = list()
+    if use_threads is False:
+        dfs = list(
+            _execute_query(
+                query_id=query_id,
+                token_work_unit=token_work_unit,
+                categories=categories,
+                safe=safe,
+                use_threads=use_threads,
+                boto3_session=boto3_session,
+            )
+            for token_work_unit in token_work_units
+        )
+    else:
+        cpus: int = _utils.ensure_cpu_count(use_threads=use_threads)
+        with concurrent.futures.ThreadPoolExecutor(max_workers=cpus) as executor:
+            dfs = list(
+                executor.map(
+                    _execute_query,
+                    itertools.repeat(query_id),
+                    token_work_units,
+                    itertools.repeat(categories),
+                    itertools.repeat(safe),
+                    itertools.repeat(use_threads),
+                    itertools.repeat(_utils.boto3_to_primitives(boto3_session=boto3_session)),
+                )
+            )
+    dfs = [df for df in dfs if not df.empty]
+    if not chunked:
```

Comment:
This implementation of the `chunked` flag will not produce a generator like described in the `Returns` section of the docstring of the `read_sql_query()` function.

We should keep that behavior and update the doc or update the behavior and keep the doc.

I would prefer to change the behavior because the current one doesn't help the user too much.
The mainly idea behind the `chunked` flag is to ingest one data chunk at a time provisioning only the amount of memory necessary for each chunk (in a lazy style).
But currently we are ingesting all chunks at once and returning a list of chunks provisioning the same amount of memory consumed by `chunked=False`.

What do you think?

_P.S._ We can even drop the `chunked` feature for our first release if you think it is better.

_P.P.S._ I think the `chunked` implementation would be similar to the `if use_threads is False:` case, but with a `for` instead of the list comprehension and a `yield` to return each chunk inside the `for` loop.

Reply:
I agree, I think we have enough time before release to implement it so will reach out to you to discuss it